### PR TITLE
Feature/vaa 5

### PR DESCRIPTION
**Pull request by vitaly-lab**

### CODE REVIEW

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+    private static final String INSERT_AIRPLANE_SQL = "INSERT INTO airplanes(code_name, model, manufacture, capacity, flight, crew_id) VALUES (?, ?, ?, ?, ?, ?);";
+    private final static String SELECT_AIRPLANE_BY_ID_SQL = "SELECT * FROM airplanes WHERE airplanes.code_name = ?;";
+    private final static String SELECT_ALL_AIRPLANE_SQL = "SELECT * FROM airplanes;";
+    private static final String DELETE_AIRPLANE_SQL = "delete from airplanes where id = ?;";
+    private static final String SELECT_CREW_MEMBERS_BY_CREW_NAME = "SELECT * FROM airplanes LEFT JOIN crews " +
+            "ON airplanes.crews_id = crews.id WHERE crews.name = ?";
+    private static final String UPDATE_AIRPLANE_CREWS_SQL = "UPDATE airplanes SET crew_id = ? WHERE id = ?;";
+
+    private DataSource dataSource;
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: could be final

Comment (joseph-dredd) [reply src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: where it supposed to initialize? I guess it should be constructor

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            return fillStatementWithAccountData(insertStatement, airplane);
+        } catch (SQLException e) {
+            throw new DaoOperationException("Cannot prepare statement to insert Airplane", e);
+        }
+    }
+
+    private PreparedStatement fillStatementWithAccountData(PreparedStatement insertStatement, Airplane airplane)
+            throws SQLException {
+        insertStatement.setString(1, airplane.getCodeName());
+        insertStatement.setString(2, airplane.getCodeName());
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/AirplaneDaoImpl.java:62]: please carefully check:
insertStatement.setString(1, airplane.getCodeName());
insertStatement.setString(2, airplane.getCodeName());
and :
   insertStatement.setInt(5, airplane.getCapacity());
        insertStatement.setInt(5, airplane.getFlightRange());
        insertStatement.setInt(5, airplane.getCrews_id());
        
        
        when 5, 5 and 5 duplicating, so:
        
        check all numbers: should be unique, 
        check all fieldNames, i.e. airplane.getCrews_id, should be getCrewId

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            return selectByIdStatement;
+        } catch (SQLException e) {
+            throw new DaoOperationException("Cannot prepare statement to select airplane by codeName", e);
+        }
+    }
+
+    private Airplane parseRow(ResultSet resultSet) throws SQLException {
+        return Airplane.builder()
+                .withId(resultSet.getInt("id"))
+                .withCodeName(resultSet.getString("codeName"))
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: compare with https://github.com/vitaly-lab/airlines-app/blob/feature/VAA-4/src/main/java/com/airlines/dao/CrewDaoImpl.java

code_name instead codeName, resultSet contains database table row

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    @Override
+    public List<Airplane> findAll() {
+        try (Connection connection = dataSource.getConnection()) {
+            Statement statement = connection.createStatement();
+            ResultSet resultSet = statement.executeQuery(SELECT_ALL_AIRPLANE_SQL);
+            return collectToList(resultSet);
+        } catch (SQLException e) {
+            throw new DaoOperationException(e.getMessage());
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: it's better throw custom exception:
throw new DaoOperationException("Can not perform find all operation", e));

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        return accountList;
+    }
+
+    public boolean deleteAirplane(int id) {
+        boolean rowDeleted;
+        try (Connection connection = dataSource.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(DELETE_AIRPLANE_SQL);
+            statement.setInt(1, id);
+            rowDeleted = statement.executeUpdate() > 0;
+            return rowDeleted;
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: try to inline variables:
statement.setInt(1, id);
return statement.executeUpdate() > 0;

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        } catch (SQLException e) {
+            throw new DaoOperationException(e.getMessage());
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/AirplaneDaoImpl.java:152]: try to point out custom message format("Can not delete airplane with id = %s", s), e)

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+            throw new DaoOperationException("Can not perform select Airplane by id query", e);
+        }
+    }
+
+    public void updateAirplane(Airplane airplane, int crews_id) {
+        try (Connection connection = dataSource.getConnection()) {
+            PreparedStatement updateStatement = prepareUpdateStatement(airplane, crews_id, connection);
+            executeUpdate(updateStatement, "Airplane was not updated");
+        } catch (SQLException e) {
+            throw new DaoOperationException(e.getMessage());
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: try to use  ("Can not update airplane with id = %s", s), e)

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+                Airplane airplane = parseRow(resultSet);
+                list.add(airplane);
+            }
+            return list;
+        } catch (SQLException e) {
+            throw new DaoOperationException("Can not perform select Airplane by id query", e);
+        }
+    }
+
+    public void updateAirplane(Airplane airplane, int crews_id) {
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: crews_id - violation Java naming convention

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    private PreparedStatement prepareUpdateStatement(Airplane airplane, int id, Connection connection) {
+        try {
+            PreparedStatement updateStatement = connection.prepareStatement(UPDATE_AIRPLANE_CREWS_SQL);
+            fillStatementWithAccountData(updateStatement, airplane);
+            updateStatement.setInt(1, id);
+            return updateStatement;
+        } catch (SQLException e) {
+            throw new DaoOperationException(e.getMessage());
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: throw new DaoOperationException("Cannot prepare statement to update Airplane", e);

File: src/main/java/com/airlines/dao/AirplaneDao.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+    void save(Airplane airplane);
+
+    Airplane findOne(String codeName);
+
+    List<Airplane> findAll();
+
+    boolean deleteAirplane(int id);
+
+    List<Airplane> searchAirplanes(String nameCrew);
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/AirplaneDao.java]: crewName

File: src/main/java/com/airlines/dao/AirplaneDao.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+    void updateAirplane(Airplane airplane, int crews_id);
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/AirplaneDao.java:19]: crewId

**Review by joseph-dredd — CHANGES_REQUESTED** [review]

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java (hunk 1 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: fixed in [3d2fa27](https://github.com/vitaly-lab/airlines-app/pull/6/commits/3d2fa27df397dda1795b8b92332e2576be3067c6)

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java (hunk 2 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/AirplaneDaoImpl.java:62]: fixed in [3d2fa27](https://github.com/vitaly-lab/airlines-app/pull/6/commits/3d2fa27df397dda1795b8b92332e2576be3067c6)

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java (hunk 3 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: fixed in [3d2fa27](https://github.com/vitaly-lab/airlines-app/pull/6/commits/3d2fa27df397dda1795b8b92332e2576be3067c6)

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java (hunk 4 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: fixed in [3d2fa27](https://github.com/vitaly-lab/airlines-app/pull/6/commits/3d2fa27df397dda1795b8b92332e2576be3067c6)

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java (hunk 5 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: fixed in [3d2fa27](https://github.com/vitaly-lab/airlines-app/pull/6/commits/3d2fa27df397dda1795b8b92332e2576be3067c6)

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java (hunk 7 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: fixed in [3d2fa27](https://github.com/vitaly-lab/airlines-app/pull/6/commits/3d2fa27df397dda1795b8b92332e2576be3067c6)

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java (hunk 8 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: fixed in [3d2fa27](https://github.com/vitaly-lab/airlines-app/pull/6/commits/3d2fa27df397dda1795b8b92332e2576be3067c6)

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java (hunk 9 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: fixed in [3d2fa27](https://github.com/vitaly-lab/airlines-app/pull/6/commits/3d2fa27df397dda1795b8b92332e2576be3067c6)

File: src/main/java/com/airlines/dao/AirplaneDao.java (hunk 10 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/AirplaneDao.java]: fixed in [3d2fa27](https://github.com/vitaly-lab/airlines-app/pull/6/commits/3d2fa27df397dda1795b8b92332e2576be3067c6)

File: src/main/java/com/airlines/dao/AirplaneDao.java (hunk 11 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/AirplaneDao.java:19]: fixed in [3d2fa27](https://github.com/vitaly-lab/airlines-app/pull/6/commits/3d2fa27df397dda1795b8b92332e2576be3067c6)

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+    private Airplane parseRow(ResultSet resultSet) throws SQLException {
+        return Airplane.builder()
+                .withId(resultSet.getInt("id"))
+                .withCodeName(resultSet.getString("code_name"))
+                .withModel(Model.valueOf(resultSet.getString("model")))
+                .withManufacture(resultSet.getDate("manufactureDate").toLocalDate())
+                .withCapacity(resultSet.getInt("capacity"))
+                .withFlightRange(resultSet.getInt("flightRange"))
+                .withCrewId(resultSet.getInt("crewId"))
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: please check column name. Should be crew_id. Database column names follows under_score_convention.

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+            throw new DaoOperationException("Cannot prepare statement to select airplane by codeName", e);
+        }
+    }
+
+    private Airplane parseRow(ResultSet resultSet) throws SQLException {
+        return Airplane.builder()
+                .withId(resultSet.getInt("id"))
+                .withCodeName(resultSet.getString("code_name"))
+                .withModel(Model.valueOf(resultSet.getString("model")))
+                .withManufacture(resultSet.getDate("manufactureDate").toLocalDate())
```
Comment (joseph-dredd) [inline src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: do you have manufactureDate column in database?

**Review by joseph-dredd — CHANGES_REQUESTED** [review]

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java (hunk 12 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: fixed in [3d2fa27](https://github.com/vitaly-lab/airlines-app/pull/6/commits/3d2fa27df397dda1795b8b92332e2576be3067c6)

File: src/main/java/com/airlines/dao/AirplaneDaoImpl.java (hunk 13 above)
Comment (vitaly-lab) [reply src/main/java/com/airlines/dao/AirplaneDaoImpl.java]: ixed in [3d2fa27](https://github.com/vitaly-lab/airlines-app/pull/6/commits/3d2fa27df397dda1795b8b92332e2576be3067c6)

**Pull request merged**